### ISSUE TITLE
Error in MFI calculation

### ISSUE THREAD

**Issue by UnicornsExist** [bukosabino/ta#54]
In your documentation, you have a link to MFI formula explanation and there you can see such text: "Money flow is positive when the typical price rises (buying pressure) and negative when the typical price declines (selling pressure)." So, you need to compare typical prices, not close prices.

    df.loc[(df['Close'] > df['Close'].shift(1, fill_value=df['Close'].mean())), 'Up_or_Down'] = 1
    df.loc[(df['Close'] < df['Close'].shift(1, fill_value=df['Close'].mean())), 'Up_or_Down'] = 2

Here you need to make a comparison with tp.

**Comment by bukosabino** [comment]
These lines are to split negative and positive values.

It compares with `typical price` below.

Best,
Dario

**Comment by zachjgal** [comment]
Hey, I think this is a legitimate bug. It seems what UnicornsExist is getting at is that when you split the negative and positive values, you do that based on the typical price, not the closing price. So you have to calculate the typical prices first and use those to decide which money flows are positive and negative,

Either way, the function doesn't give correct values. I unit tested it with an existing dataset for which MFI was already calculated, and it failed. I thought perhaps the dataset used for the test was wrong, so I wrote my own function to calculate the MFI, but it agreed with the test dataset.

**Comment by UnicornsExist** [comment]
zachjgal, yes, exactly what I mean

**Comment by StefanoCore** [comment]
@zachjgal Can you please share the MSI function you built by yourself so we can test it against the one on the library?

**Comment by zachjgal** [comment]
@StefanoCore Sure thing. It was very much written on the fly for a project, so it's poorly optimized and features some ugly-looking pandas,  but it's fast enough for my purposes and you can see all the steps of the calculation.

```
def mfi(high, low, close, volume, n=14):
    """
    Calculates the money flow index, or MFI for short. Note that any series
    specified in args are expected to be a time series of daily values, indexed by date.

    The MFI is a way of measuring "enthusiasm" in the market, and can
    identify stocks that may be overbought or oversold. It's often compared
    to the relative strength index (RSI) indicator, but incorporates
    additional information about a stock's volume.

    Args:
        high (pd.Series): The highest price of the stock for each trading day.
        low (pd.Series): The lowest price of the stock for each trading day,
        close (pd.Series): The price of the stock when the market closed on
            each day
        volume (pd.Series): The total number of shares that changed hands for
            each trading day.
        n (int): The lookback period, or the number of trading days for which
            past data should be used. The standard is 14.

    Returns:
        pd.Series: The MFI for each day
    """
    df = pd.concat([high, low, close, volume], axis=1)
    df.columns = ['High', 'Low', 'Close', 'Volume']
    df['Date'] = df.index
    df.index = list(range(len(df.index)))
    # typical price
    df['Typ'] = (df['High'] + df['Low'] + df['Close']) / 3
    # raw money flow
    df['MF'] = df['Typ'] * df['Volume']
    df['MF_sign'] = df['Typ'].diff()
    # money ratio and MFI
    df['MR'] = np.nan
    mr_loc = df.columns.get_loc('MR')
    df['MFI'] = np.nan
    mfi_loc = df.columns.get_loc('MFI')
    for index, row in df.iloc[n-1:, :].iterrows():
        data_to_use = df.iloc[index - n + 1: index + 1]
        positive_flow = data_to_use[data_to_use['MF_sign'] >= 0]['MF']
        negative_flow = data_to_use[data_to_use['MF_sign'] < 0]['MF']
        total_positive_flow = sum(positive_flow.values)
        total_negative_flow = sum(negative_flow.values)
        # prevents rare issue with dividing by 0
        if total_negative_flow == 0:
            money_ratio = 1000
        else:
            money_ratio = total_positive_flow/total_negative_flow
        df.iat[index, mr_loc] = money_ratio
        df.iat[index, mfi_loc] = 100 - (100 / (1 + money_ratio))
    df.index = df['Date']
    
    return df['MFI']
```
Here's a link to an excel file of the sample data I tested on:
https://drive.google.com/file/d/1WNj_Cw4fw1eZsyDdyqIiFgAgz_W6bVXI/view?usp=sharing

**Comment by bukosabino** [comment]
Hi,

I have implemented a new version fixing the bug. It passes the test provided by @zachjgal 

I will include in the next ta library version (next days).

Best,
Dario

**Comment by bukosabino** [comment]
This implementation was included today in the new 0.4.6 version.

```sh
pip install --upgrade ta
```

Thank you to all of you!
